src/Microsoft.OData.Core/ODataStreamItem.cs

```diff
@@ -0,0 +1,81 @@
+//---------------------------------------------------------------------
+// <copyright file="ODataStreamItem.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+namespace Microsoft.OData
+{
+    #region Namespaces
+    using Edm;
+    #endregion
+
+    /// <summary>
+    /// A class to represent a stream value
+    /// </summary>
+    public sealed class ODataStreamItem : ODataItem
+    {
+        /// <summary>Primitive Type, if known.  Either String, Binary, or None.</summary>
+        private EdmPrimitiveTypeKind typeKind;
+
+        /// <summary>
+        /// Creates an ODataStreamValue with a specified typeKind.
+        /// </summary>
+        /// <param name="typeKind">The type of stream value.</param>
+        public ODataStreamItem(EdmPrimitiveTypeKind typeKind)
+        {
+            this.PrimitiveTypeKind = typeKind;
+        }
+
+        /// <summary>
+        /// Creates an ODataStreamValue with a specified typeKind.
+        /// </summary>
+        /// <param name="primitiveTypeKind">PrimitiveTypeKind of the stream value.</param>
+        /// <param name="contentType">The mime type of the stream value.</param>
+        /// <remarks>primitiveTypeKind is overridden if the MimeType is known.</remarks>
+        public ODataStreamItem(EdmPrimitiveTypeKind primitiveTypeKind, string contentType)
+        {
+            this.PrimitiveTypeKind = primitiveTypeKind;
+            this.ContentType = contentType;
+        }
+
+        /// <summary>
+        /// PrimitiveTypeKind of the stream.
+        /// </summary>
+        /// <returns>
+        /// EdmPrimitiveTypeKind.String, if the contents is a plain text string;
+        /// EdmPrimitiveTypeKind.Binary, if the contents is binary;
+        /// EdmPrimitiveTypeKind.None is returned if the contents could be binary or plain text
+        /// </returns>
+        public EdmPrimitiveTypeKind PrimitiveTypeKind
+        {
+            get
+            {
+                return this.typeKind;
+            }
+
+            set
+            {
+                if (typeKind != EdmPrimitiveTypeKind.String &&
+                    typeKind != EdmPrimitiveTypeKind.Binary &&
+                    typeKind != EdmPrimitiveTypeKind.None)
+                {
+                    throw new ODataException(Strings.StreamItemInvalidPrimitiveKind(value));
+                }
+
+                this.typeKind = value;
+            }

```

Comment:
why do we allow to set the property which is set in the constructor?

Reply:
Excellent question.

Originally I made it settable only in the constructor, but then I added ContentType. I initially had one constructor that took PrimitiveTypeKind and a second that took just ContentType, and set the primitivetypekind based on the content type. But that was just a guess, and could be wrong, so then I allowed explicitly setting the content type.  But then I decided just to pass both primitivetypekind and contenttype to the second constructor, so it was no longer necessary to explicitly set either the content type or primitive type, so I've now made both setters private.

---
In reply to: [268358264](https://github.com/OData/odata.net/pull/1204#discussion_r268358264) [](ancestors = 268358264)